### ISSUE TITLE
Ex 2 level3.jsh

### ISSUE THREAD

**Issue by rayray39** [nus-cs2030/2223-s2#203]
Hi, I have tried to remove the services from my active/expired list using a normal for loop with int i = 0... but i realised that the size of the list changes and its very easy to make an error so i decided to use the iterator method of ImList to remove the services. But my lists dont seem to be generating the correct result, for eg. when i remove the services from active, they dont get remvoved and my expired list remains empty. Any tips on how to properly remove and add the services?

**Comment by Ernestnzx** [comment]
Personally I used the `for (Cruise cruise : cruises)` to iterate through the `active` and the `expired` queue, I didn't experience any issue with this method! Just remember to assign the new ImList for both queues and generate new Loaders and add them to the `active` queue when needed.

**Comment by jaejayrome** [comment]
You could consider using a while loop also and initialise a variable outside of the while loop: eg int counter = 0; which would help to keep track of (e.g) the number of loaders that already been instructed to serve the given cruise. However, your issue looks to me that you might be forgetting to reassign the ImList to the same variable by pointing the new ImList instance returned from the add() or remove() methods for example. 

what I mean is as such: 
activeQueue = activeQueue.remove(index_of_service);

or (in the case of adding to the expired list):

expiredQueue = expiredQueue.add(service_instance);

Hope this solves your issues!

**Comment by rayray39** [comment]
> 

if i were to use the indices to remove, wouldnt the size of the lists change as you remove stuff? Or isit just fix the index_of_service = 0 for the removal of services in active? and how did you keep track of the Loader ID? any tips will be much appreciated! Thanks!!

**Comment by territxl** [comment]
I dont think so as in the ImList.java, in the method Remove, a new ImList is created with the element removed. Hence, I don't think it will affect the list since it is effect-free :)

**Comment by Yew-Wei** [comment]
Hi, I think I understand your most recent question. You could possibly iterate using a for each loop on another copy of your actual service queue and remove the 0 index of the actual service queue itself. This is done because your services should already be in chronological order, hence for the time condition to hold true when iterating through the actual service queue copy should hold true for the index 0th element. I hope it makes sense to you. 

Regarding your second point you could initialise a counter for the loader ID as mentioned above!

**Comment by sjcze** [comment]
You can use the enhanced for loop to iterate over the active list, removal can be done by calling the indexOf method from ImList.

**Comment by rshalinisai** [comment]
It might be bacause when you iterate through the activeServices and expiredServices list, you also remove elements from the lists in the for loop --> so the size of the list varies and you end up removing a different element from the one that you are intending to remove

**Comment by rshalinisai** [comment]
When I faced this issue, I created an updated list. I iterated thru the original list, then during the for loop I removed from the updated list not the original list because I don't want to change the size of the original list in the for loop. Then, after the for loop I reassigned the original list to be the updated list.

**Comment by littlemidget3** [comment]
make use of a copy of the list so u can preserve the original list, this can be applied to situations when you iterate through lists and have to update the list at the same time. Apologies for the late comment!
